pwa: Add a Progressive Web App (PWA) plugin

pwithnall opened the pull request:
<!--
 1. Following guidance below, replace …'s with your own words
 2. After saving the PR, tick of completed checklist items
 3. Skip checklist items that are not applicable or not necessary
 4. Delete instruction/comment blocks
-->

## Summary

This allows installing a Kolibri instance as a web app on the user’s desktop, if their web browser supports it (Chrome does) and if the site is configured to allow it (HTTPS must be enabled, it must be accessible without an internet connection, and the theme must have appropriate icons defined).

There are two main components to making an installable PWA: a manifest and a service worker. See https://www.w3.org/TR/appmanifest/ or https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps for an overview of PWAs.
 
We serve the manifest from `/${lang}/pwa/manifest.json`, and it’s linked to from the `<head>` of every page so that the browser can find it. It’s localised using the normal URI i18n functionality. The web app manifest specification doesn’t allow for localisation within the manifest file.

The service worker has to be served from `/sw.js` (not `/${lang}/pwa/sw.js`) because its URI path determines its scope. If it were in a nested path, it could only act on requests at that path or beneath, which wouldn’t work. See https://www.w3.org/TR/service-workers/#path-restriction.

## References

 * PWA specification: https://www.w3.org/TR/appmanifest/
 * PWA introduction: https://web.dev/add-manifest/
 * Previous PR (rejected) for adding PWA support to the core: #10713, including discussion of the approach

## Reviewer guidance

 * Test by compiling and running Kolibri as normal and running `kolibri plugin enable pwa`
 * You’ll need to test in Chrome or another web browser which supports PWAs
 * I’m new to working on Django/Kolibri, so may not have followed the right conventions
 * I’ve left the gzip compression in for now, since it reduces the size of the files by ~50% and they should be cached in that state, providing benefits across the network; can remove it if needed though
 * The `decorators` commit is not strictly needed for this version of the plugin (it’s left over from the previous iteration of the PR), but I think it’s a correct change regardless, so it’s up for review

----

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests


## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

pwithnall committed:
core: Add FrontEndBaseHeadHook for injecting markup into the <head>

This allows plugins to inject arbitrary markup into the `<head>` of
every page. This is similar to what `FrontEndBaseSyncHook` allows, but
it’s limited to linking to CSS and JS files.

This will be needed by the PWA plugin to add a `<link>` to the PWA
manifest for the site.

Signed-off-by: Philip Withnall <pwithnall@endlessos.org>

pwithnall committed:
core: Allow specifying additional logos in the theme

Add a `logos` array for defining zero or more logos to be used on the
site. These will be used in an upcoming commit for exposing the site’s
logo in various sizes in a PWA manifest.

Add additional `content_type` and `maskable` properties to the image
spec to help with this. Both are optional. `content_type` specifies the
content type of the image (such as `image/png`). `maskable` specifies
whether the image is layed out so that it can be masked by a client (see
https://web.dev/maskable-icon/).

Signed-off-by: Philip Withnall <pwithnall@endlessos.org>

pwithnall committed:
default_theme: Expose the default logo in various sizes

This provides a placeholder logo to be used by the PWA plugin (coming in
a subsequent commit), and documents how to define logos in custom
themes.

Signed-off-by: Philip Withnall <pwithnall@endlessos.org>

pwithnall committed:
pwa: Add a Progressive Web App (PWA) plugin

This allows installing a Kolibri instance as a web app on the user’s
desktop, if their web browser supports it (Chrome does) and if the site
is configured to allow it (HTTPS must be enabled, it must be accessible
without an internet connection, and the theme must have appropriate
icons defined).

There are two main components to making an installable PWA: a manifest
and a service worker. See https://www.w3.org/TR/appmanifest/ or
https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps for an
overview of PWAs.

We serve the manifest from `/${lang}/pwa/manifest.json`, and it’s linked
to from the `<head>` of every page so that the browser can find it. It’s
localised using the normal URI i18n functionality. The web app manifest
specification doesn’t allow for localisation within the manifest file.

The service worker has to be served from `/sw.js` (not
`/${lang}/pwa/sw.js`) because its URI path determines its scope. If it
were in a nested path, it could only act on requests at that path or
beneath, which wouldn’t work. See
https://www.w3.org/TR/service-workers/#path-restriction.

Signed-off-by: Philip Withnall <pwithnall@endlessos.org>

rtibbles commented on kolibri/core/decorators.py:
```diff
@@ -373,7 +373,7 @@ def inner_func(*args, **kwargs):
         render_and_cache(response, CACHE_KEY_TEMPLATE.format(request.path))
         patch_response_headers(response, cache_timeout=CACHE_TIMEOUT)
-        response["Vary"] = "accept-encoding, accept"
+        response["Vary"] = "accept-encoding, accept, accept-language"
```
This change should be reverted - this decorator is only intended for use on pages that are already prefixed by language codes in their URL path, so we explicitly exclude accept-language to ensure higher cache hit rates.

rtibbles commented on kolibri/plugins/pwa/templates/manifest.json:
```diff
@@ -0,0 +1,37 @@
+{% load i18n core_tags webpack_tags content_tags cache %}
```
I don't think you are using `core_tags webpack_tags content_tags cache` here, so you could probably just leave this as `i18n`.

rtibbles commented on kolibri/plugins/pwa/templates/sw.js:
```diff
@@ -0,0 +1,8 @@
+
+/* Stub ServiceWorker to keep Chrome happy. */
+self.addEventListener('install', (event) => {
+    console.log('[SW] Installing ServiceWorker version:', VERSION);
```
Not important, but I think in most developer consoles, console.logs from the service worker will be prefixed anyway.

rtibbles commented on kolibri/plugins/pwa/urls.py:
```diff
@@ -0,0 +1,13 @@
+from .views import PwaManifestView
+
+urlpatterns = [
+    url(r"^manifest.json$", PwaManifestView.as_view(), name="manifest"),
```
I think we should just be using one of these, not both? The shared `name` property between these two urls will cause issues as well.

It seems that `.webmanifest` is the preferred extension:

```
NOTE: File extension: .webmanifest or .json?
The IANA [registered](https://www.iana.org/assignments/media-types/application/manifest+json) file extension for the manifest is .webmanifest. Some web servers recognize this extension and transfer the file using the standardized [application manifest media type](https://www.w3.org/TR/appmanifest/#dfn-application-manifest-media-type) ([application/manifest+json](https://www.w3.org/TR/appmanifest/#dfn-application-manifest-json)). Developers can also choose a different extension (e.g. .json) or none at all (e.g. /api/GetManifest), but are encouraged to transfer the manifest using the [application/manifest+json](https://www.w3.org/TR/appmanifest/#dfn-application-manifest-json) [MIME type](https://mimesniff.spec.whatwg.org/#mime-type), although any [JSON MIME type](https://mimesniff.spec.whatwg.org/#json-mime-type) is ok.
```

And as you are setting the content_type on the template view, this should be OK?

pwithnall replied:
OK, I’ve reduced it to `.webmanifest`. I’ve also fixed the content-type; it was previously `application/json`, but it’s now `application/manifest+json`.

rtibbles commented on kolibri/plugins/pwa/templates/sw.js:
```diff
@@ -0,0 +1,8 @@
+{% load i18n core_tags webpack_tags content_tags cache %}
```
I don't think any of these are needed here.

rtibbles commented on kolibri/plugins/pwa/kolibri_plugin.py:
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright 2023 Endless OS Foundation, LLC
```
We don't have an explicit contributor license grant in place, but we can't accept code into our code base that is under copyright to another organization, without an associated license disclaimer.

My preference would be for this copyright notice to be removed.

pwithnall replied:
As discussed on Slack, I’ve kept the copyright line and also added a one-line license identifier (in all new files).

rtibbles commented on kolibri/plugins/pwa/root_urls.py:
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright 2023 Endless OS Foundation, LLC
```
Ditto.

rtibbles commented on kolibri/plugins/pwa/urls.py:
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright 2023 Endless OS Foundation, LLC
```
Ditto.

rtibbles commented on kolibri/plugins/pwa/views.py:
```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright 2023 Endless OS Foundation, LLC
```
Ditto.

rtibbles reviewed (changes requested):
Mostly there, but a bit of cleanup needed, and some legal minutiae!

rtibbles commented on kolibri/plugins/default_theme/kolibri_plugin.py:
```diff
@@ -25,4 +25,25 @@ def theme(self):
+                    "content_type": "image/svg+xml",
+                    # See https://web.dev/maskable-icon/ for details on what
+                    # icons count as maskable. The default Kolibri logo is not,
+                    # as the outer ‘waves’ circle gets cropped.
```
I think these forward ticks are causing the non-ASCII character error in the Python 2.7 tests.

pwithnall replied:
I’ve changed them to ASCII single quotes. It’s slightly upsetting that 3 years after its EOL, Python 2.7 is still holding things back!

pwithnall committed:
authors: Add myself

For adding the Progressive Web App (PWA) plugin.

Signed-off-by: Philip Withnall <pwithnall@endlessos.org>

pwithnall commented:
Updated to address the review comments, thanks for the detailed review. I’ve also run `black` over it (which should hopefully fix the style-check CI) and have added myself to `AUTHORS.md`.

I’ll submit a further PR later to update a few bits of documentation about copyright/licensing, based on our discussion from Slack.

rtibbles commented:
Looks like the linting still isn't quite what the CI expects. Easiest way to fix it is to install `pre-commit` and run `pre-commit run --all-files` to make sure it's running with exactly the same hooks/black version etc.

There's also a weird error happening where it appears the PWA plugin is active and the hook registered during the URLs test, but it has not been enabled, so the templates are not registered to Django.

So, some shenanigans are happening whereby the `kolibri_plugin.py` has been imported and the hook registered at some point during a test somewhere, but the plugin isn't enabled.

pwithnall committed:
utils: Don’t initialise hooks when iterating plugins

The PWA plugin registers a hook which is used on every page. The
`test_key_urls` test opens all the URLs in the main URL map, which
includes `/plugins/` and then `/learn/`. If iterating plugins
initialises hooks, and the PWA plugin is disabled, then loading
`/plugins/` causes the PWA’s hook to be registered, and then loading
`/learn/` causes it to be executed — but the plugin is not enabled. This
causes an exception.

This doesn’t change the behaviour of `kolibri plugin list` or the
`/plugins/` endpoint on a normal installation.

Signed-off-by: Philip Withnall <pwithnall@endlessos.org>

pwithnall commented:
> Looks like the linting still isn't quite what the CI expects. Easiest way to fix it is to install `pre-commit` and run `pre-commit run --all-files` to make sure it's running with exactly the same hooks/black version etc.

I couldn’t get `pre-commit` working and didn’t dig into the failure, so have manually fixed the linting problems highlighted in the previous CI test run. Hopefully that’s sufficient.

> There's also a weird error happening where it appears the PWA plugin is active and the hook registered during the URLs test, but it has not been enabled, so the templates are not registered to Django.
> 
> So, some shenanigans are happening whereby the `kolibri_plugin.py` has been imported and the hook registered at some point during a test somewhere, but the plugin isn't enabled.

I think I’ve found the cause of this, and have fixed it in https://github.com/learningequality/kolibri/pull/10723/commits/0135befbf0d33f58fdeca693b87b1c8cb5f3948b. The tests pass locally for me with `tox` now.

rtibbles commented:
Thanks - that does indeed look like the right fix!

pwithnall commented:
Looks like I missed one linting warning in `hooks.py`. Just pushed another update to fix that (no other changes).

rtibbles commented:
I do keep planning on integrating the pre-commit autofixing for PRs, but haven't got around to it yet!

pwithnall commented:
Thanks for the patient and helpful reviews!
